### Import all the things

In [130]:
# special IPython command to prepare the notebook for matplotlib
%matplotlib inline 

import numpy as np
import pandas as pd
import scipy.stats as stats
import matplotlib.pyplot as plt
import sklearn
import statsmodels.api as sm

import seaborn as sns
sns.set_style("whitegrid")
sns.set_context("poster")

# special matplotlib argument for improved plots
from matplotlib import rcParams

Load the parts cost data

In [131]:
# assign column names for each CSV
l_cols = ['id_partL1', 'num_level', 'id_tparent', 'desc_uom', 'id_task', 'desc_status', 'id_part', 'num_totLead', 'num_dimLead', 
          'num_inv', 'num_postProdLead', 'num_preProdLead', 'num_procLead', 'num_itemMatCost', 'num_stockInd', 'num_onHand', 
          'num_resQty', 'num_transQty', 'num_bomQty', 'num_poCost', 'num_poQty', 'num_itemCost', 'num_onHand', 'num_onHandInv', 
          'num_openPOsTot', 'num_openSOsTot', 'num_soQty', 'num_soUnitCost', 'num_venLead', 'num_venPOQty', 'num_venPOPrice' ]

parts_cost = pd.read_csv(
    'C:/Users/502689880/GE Projects/Lead_Times/ltData.csv',
    names=l_cols)

parts_cost.head()

id_partL1  num_level    id_tparent   desc_uom id_task desc_status  \
0  143E7165G001          4  367A6377G007  EA (each)    0010  Superceded   
1  143E7165G001          4  367A6377G007  EA (each)    0010  Superceded   
2  143E7165G001          4  367A6377G007  EA (each)    0010  Superceded   
3  144E7446G001          4  367A6571G001  EA (each)    0010  Superceded   
4  144E7446G001          4  367A6571G001  EA (each)    0010  Superceded   

               id_part  num_totLead  num_dimLead  num_inv       ...        \
0  X0006AAX0Y0BA030-0X   108.571429          0.0      0.0       ...         
1    X0005A0M0BA050-0J   110.000000          0.0      0.0       ...         
2    X0005A0M0BA060-0J    53.371429          0.0      0.0       ...         
3    X0005CAG0BA020-0D   108.571429          0.0      0.0       ...         
4    X0005CAG0BA020-0E   108.571429          0.0      0.0       ...         

   num_itemCost  num_onHand.1  num_onHandInv  num_openPOsTot  num_openSOsTot  \
0           0.0           0.0              0               0               0   
1           0.0           0.0              0               0               0   
2           0.0           0.0              0               0               0   
3           0.0           0.0              0               0               0   
4           0.0           0.0              0               0               0   

   num_soQty  num_soUnitCost  num_venLead  num_venPOQty  num_venPOPrice  
0        0.0             0.0          0.0           0.0             0.0  
1        0.0             0.0          0.0           0.0             0.0  
2        0.0             0.0          0.0           0.0             0.0  
3        0.0             0.0          0.0           0.0             0.0  
4        0.0             0.0          0.0           0.0             0.0  

[5 rows x 31 columns]

### The basics

In [132]:
parts_cost.dtypes

id_partL1            object
num_level             int64
id_tparent           object
desc_uom             object
id_task              object
desc_status          object
id_part              object
num_totLead         float64
num_dimLead         float64
num_inv             float64
num_postProdLead    float64
num_preProdLead     float64
num_procLead        float64
num_itemMatCost     float64
num_stockInd          int64
num_onHand          float64
num_resQty          float64
num_transQty        float64
num_bomQty          float64
num_poCost          float64
num_poQty           float64
num_itemCost        float64
num_onHand.1        float64
num_onHandInv         int64
num_openPOsTot        int64
num_openSOsTot        int64
num_soQty           float64
num_soUnitCost      float64
num_venLead         float64
num_venPOQty        float64
num_venPOPrice      float64
dtype: object

The list of data types is truncated and we want to see them all. So, we'll create a series of the columns and group them by data types.

In [133]:
parts_cost.shape

(159910, 31)

159910 rows times 31 columns. A spredsheet is a table is a matrix. How can we access members of this tuple (brackets like so:() )

These are the column names.

In [134]:
parts_cost.columns

Index(['id_partL1', 'num_level', 'id_tparent', 'desc_uom', 'id_task',
       'desc_status', 'id_part', 'num_totLead', 'num_dimLead', 'num_inv',
       'num_postProdLead', 'num_preProdLead', 'num_procLead',
       'num_itemMatCost', 'num_stockInd', 'num_onHand', 'num_resQty',
       'num_transQty', 'num_bomQty', 'num_poCost', 'num_poQty', 'num_itemCost',
       'num_onHand.1', 'num_onHandInv', 'num_openPOsTot', 'num_openSOsTot',
       'num_soQty', 'num_soUnitCost', 'num_venLead', 'num_venPOQty',
       'num_venPOPrice'],
      dtype='object')

In [135]:
parts_cost.describe()

num_level    num_totLead    num_dimLead       num_inv  \
count  159910.000000  159910.000000  159910.000000  1.599100e+05   
mean        2.713783      59.354656     212.819857  1.754199e+04   
std         1.403746      43.685163     209.513802  1.152245e+05   
min         1.000000       4.666667    -276.000000  0.000000e+00   
25%         2.000000      31.000000      93.857143  1.307300e+00   
50%         3.000000      48.142857     153.889952  1.356320e+01   
75%         3.000000      71.000000     290.818792  2.800105e+02   
max        11.000000     444.300000    5104.000000  1.546020e+07   

       num_postProdLead  num_preProdLead   num_procLead  num_itemMatCost  \
count     159910.000000    159910.000000  159910.000000     1.599100e+05   
mean           8.662244         1.001017      49.691395     2.519644e+04   
std            1.944370         0.058612      43.311056     1.597862e+05   
min            0.000000         0.571429       0.000000     0.000000e+00   
25%            7.142857         1.000000      20.000000     1.330000e+00   
50%           10.000000         1.000000      39.000000     1.453333e+01   
75%           10.000000         1.000000      60.000000     3.250833e+02   
max           20.000000         9.428571     435.800000     1.919854e+07   

        num_stockInd     num_onHand       ...         num_itemCost  \
count  159910.000000  159910.000000       ...        159910.000000   
mean        0.663605     599.525368       ...            12.765906   
std         0.472477    3585.217104       ...           441.268604   
min         0.000000       0.000000       ...             0.000000   
25%         0.000000       0.000000       ...             0.000000   
50%         1.000000       9.000000       ...             0.000000   
75%         1.000000     169.000000       ...             0.000000   
max         1.000000   75203.666670       ...         50234.930000   

        num_onHand.1  num_onHandInv  num_openPOsTot  num_openSOsTot  \
count  159910.000000  159910.000000   159910.000000   159910.000000   
mean      485.753684      46.743243       14.147370       30.380439   
std      7432.758250     640.539576      297.725358      340.922017   
min         0.000000       0.000000        0.000000        0.000000   
25%         0.000000       0.000000        0.000000        0.000000   
50%         0.000000       0.000000        0.000000        0.000000   
75%         0.000000       0.000000        0.000000        0.000000   
max    354555.970000   13704.000000    19130.000000     6390.000000   

           num_soQty  num_soUnitCost    num_venLead   num_venPOQty  \
count  159910.000000    1.599100e+05  159910.000000  159910.000000   
mean       19.969332    6.474295e+12      38.580226      50.110212   
std       102.401748    2.396197e+13     139.747762     218.033391   
min         0.000000    0.000000e+00       0.000000       0.000000   
25%         0.904028    7.718229e+00      16.878788       1.125000   
50%         3.074766    8.544845e+02      24.400000       5.222222   
75%        13.097239    6.610000e+12      40.000000      24.921182   
max      5000.000000    1.000000e+15   13573.230000    6549.018519   

       num_venPOPrice  
count    1.599100e+05  
mean     3.451380e+04  
std      2.055053e+05  
min      0.000000e+00  
25%      3.538125e+00  
50%      2.265655e+01  
75%      4.469009e+02  
max      8.865269e+06  

[8 rows x 25 columns]

In [136]:
text_file = open("C:/Users/502689880/GE Projects/Parts_Cost/parts_list.txt", "r")
parts_list = text_file.read().split(',')
print (parts_list)
print (len(parts_list))
text_file.close()

parts_cost = parts_cost[parts_cost['id_partL1'].isin(parts_list)]
parts_cost = parts_cost[parts_cost['id_part'].isin(parts_list)]
parts_cost.id_partL1

['100T0494G001', '100T0911G001', '100T1067G003', '100T1067G005', '100T1168G001', '100T1511G001', '100T1657G002', '100T1657G003', '100T1657G008', '100T1657G015', '100T2215G001', '100T2380G003', '100T2468G005', '100T3075G002', '100T3075G008', '100T3079G005', '100T3311G001', '100T3605G0004', '100T3605G0005', '100T3606G0006', '100T3607G0005', '100T3607G004', '100T3904P001', '100T4144G002', '100T4149G003', '100T5421G004', '100T5950G003', '100T6006G001', '100T6370G005', '100T6370G011', '100T6851G009', '100T6852G011', '100T7631G0001', '100T7631G0002', '100T7641P002', '100T7775G001', '100T7776G016', '100T8193G003', '100T8193G004', '100T8193G006', '100T8376G011', '100T8378G008', '100T8379G012', '100T8757G001', '100T9020G007', '101L2020G004', '101T0928G001', '101T0928G002', '101T0928G006', '101T0928G007', '101T1005P002', '101T1105G001', '101T1107G019', '101T1200G001', '101T1327G001', '101T1327G003', '101T1330G0001', '101T1748G001', '101T1952G001', '101T2065G001', '101T2521G001', '101T2554G001', 

60          113T2424G011
61         108T4211G0001
64          113T2424G011
71         108T4211G0001
72         108T4211G0001
73         108T4211G0001
74         108T4211G0001
75         108T4211G0001
76         108T4211G0001
77         108T4211G0001
78         108T4211G0001
156        109T2825G0001
157        109T2825G0001
158        109T2825G0001
475        113T4984G0001
516        113T8336G0001
517        113T8336G0001
518        113T8336G0001
519        113T8336G0001
566         207C3748G001
574       336A4937BKG036
621         334A4470P002
622         334A4470P002
623        143D6518G0014
632         141E7454G008
634         357B3716G004
635         357B3716G004
647         357B3716G004
648         357B3716G004
657         357B3716G004
               ...      
142150     108T4983G0001
142151     108T4983G0001
142152     108T4983G0001
142153     108T4983G0001
142154     115T9778G0001
142155      104T4955G003
142156      104T4955G003
142157      104T4955G003
142158      143E3076G001


In [137]:
parts_cost.num_itemMatCost

60        1.782100e+04
61        1.782100e+04
64        1.782100e+04
71        1.782100e+04
72        1.782100e+04
73        1.782100e+04
74        1.782100e+04
75        1.782100e+04
76        1.782100e+04
77        1.782100e+04
78        1.782100e+04
156       2.062600e+04
157       2.062600e+04
158       2.062600e+04
475       6.175000e+03
516       1.581200e+04
517       1.581200e+04
518       1.581200e+04
519       1.581200e+04
566       0.000000e+00
574       9.291600e+00
621       1.520500e+04
622       1.520500e+04
623       6.710720e+03
632       4.261436e+04
634       2.299500e+04
635       2.299500e+04
647       2.299500e+04
648       2.299500e+04
657       2.299500e+04
              ...     
142150    0.000000e+00
142151    0.000000e+00
142152    0.000000e+00
142153    0.000000e+00
142154    0.000000e+00
142155    0.000000e+00
142156    0.000000e+00
142157    0.000000e+00
142158    0.000000e+00
142159    0.000000e+00
142160    0.000000e+00
142161    0.000000e+00
142162    0

In [138]:
parts_costsub = parts_cost[parts_cost['num_level'] == 1]
print (parts_costsub.num_level)

566       1
574       1
621       1
622       1
623       1
632       1
634       1
635       1
647       1
648       1
657       1
661       1
684       1
685       1
686       1
687       1
688       1
689       1
690       1
691       1
692       1
693       1
694       1
695       1
696       1
697       1
698       1
699       1
700       1
701       1
         ..
142114    1
142115    1
142117    1
142118    1
142149    1
142150    1
142151    1
142152    1
142153    1
142154    1
142155    1
142156    1
142157    1
142158    1
142159    1
142160    1
142161    1
142162    1
142163    1
142164    1
142165    1
142166    1
142167    1
153489    1
153492    1
153493    1
153494    1
153501    1
153502    1
153503    1
Name: num_level, Length: 8673, dtype: int64


In [139]:
parts_cost_subset = ['id_partL1', 'num_level', 'id_part', 'num_itemMatCost', 'num_itemCost', 'num_inv', 'num_poCost']
parts_cost_est = parts_costsub[parts_cost_subset]
parts_cost_est = parts_cost_est.drop_duplicates()
parts_cost_est.head()

id_partL1  num_level         id_part  num_itemMatCost  num_itemCost  \
566    207C3748G001          1    207C3748G001           0.0000           0.0   
574  336A4937BKG036          1  336A4937BKG036           9.2916           0.0   
621    334A4470P002          1    334A4470P002       15205.0000           0.0   
623   143D6518G0014          1   143D6518G0014        6710.7200           0.0   
632    141E7454G008          1    141E7454G008       42614.3600           0.0   

          num_inv  num_poCost  
566      0.000000     0.00000  
574      9.352900    63.56925  
621   8745.916000     2.00000  
623   3860.006171     0.00000  
632  24511.779890     0.00000

In [140]:
parts_cost_est

id_partL1  num_level         id_part  num_itemMatCost  \
566       207C3748G001          1    207C3748G001     0.000000e+00   
574     336A4937BKG036          1  336A4937BKG036     9.291600e+00   
621       334A4470P002          1    334A4470P002     1.520500e+04   
623      143D6518G0014          1   143D6518G0014     6.710720e+03   
632       141E7454G008          1    141E7454G008     4.261436e+04   
634       357B3716G004          1    357B3716G004     2.299500e+04   
684       116E4414G007          1    116E4414G007     2.754722e+04   
739      131E4436G0008          1   131E4436G0008     0.000000e+00   
741      112E1409G0029          1   112E1409G0029     1.071400e+04   
755      131E4436G0008          1   131E4436G0008     3.485223e+03   
766      106T7757G0001          1   106T7757G0001     3.621954e+04   
767       101T5789G001          1    101T5789G001     5.803200e+03   
795       146E3629G008          1    146E3629G008     4.748712e+03   
809       116E1215G010          1    116E1215G010     1.706528e+04   
811       101T5789G001          1    101T5789G001     0.000000e+00   
836      106T2789G0001          1   106T2789G0001     1.600000e+04   
837       100T1657G002          1    100T1657G002     1.363875e+03   
847      113T1575G0004          1   113T1575G0004     1.420000e+04   
853      113T1575G0005          1   113T1575G0005     1.420000e+04   
870       100T1657G002          1    100T1657G002     0.000000e+00   
903       100T3311G001          1    100T3311G001     4.700000e+02   
963       342A1784P001          1    342A1784P001     1.413932e+04   
970       119E3015G001          1    119E3015G001     2.085000e+03   
1085      197C3666G004          1    197C3666G004     1.675000e+03   
1089      117E3816G010          1    117E3816G010     2.030830e+05   
1092     117T2109G0001          1   117T2109G0001     2.289700e+05   
1098     113T8680G0001          1   113T8680G0001     6.413000e+01   
1099     116T6101G0001          1   116T6101G0001     6.992000e+01   
1100     116T8930G0001          1   116T8930G0001     7.900000e+03   
1101      146E4860G002          1    146E4860G002     6.912500e+04   
...                ...        ...             ...              ...   
133924    143E3937G001          1    143E3937G001     0.000000e+00   
133940   125T3381G0001          1   125T3381G0001     0.000000e+00   
133941    100T8193G006          1    100T8193G006     3.362335e+06   
133944   100T3605G0004          1   100T3605G0004     0.000000e+00   
133953   100T3606G0006          1   100T3606G0006     0.000000e+00   
133959   107T0953G0002          1   107T0953G0002     0.000000e+00   
133962    117E5633G014          1    117E5633G014     3.002819e+04   
133974    106L2290G011          1    106L2290G011     0.000000e+00   
140904    143E3137G026          1    143E3137G026     3.673786e+04   
141083    143E6721G007          1    143E6721G007     0.000000e+00   
142025    361A2225G025          1    361A2225G025     0.000000e+00   
142028   112T8358G0001          1   112T8358G0001     0.000000e+00   
142029   107T7934G0001          1   107T7934G0001     3.147490e+03   
142031   110T3421P0002          1   110T3421P0002     0.000000e+00   
142039    119E9302G004          1    119E9302G004     0.000000e+00   
142045    100T2380G003          1    100T2380G003     8.057474e+04   
142066    166D1533G010          1    166D1533G010     3.993717e+03   
142069   118T4113G0001          1   118T4113G0001     0.000000e+00   
142091   108T1854G0001          1   108T1854G0001     0.000000e+00   
142093   108T6828P0001          1   108T6828P0001     0.000000e+00   
142098   121T9411G0101          1   121T9411G0101     0.000000e+00   
142102   118T5943G0101          1   118T5943G0101     0.000000e+00   
142110    141E4665G005          1    141E4665G005     1.677591e+06   
142114    100T8379G012          1    100T8379G012     0.000000e+00   
142117    188C9517G015          1    188C9517G015     0.000000e+00   
142149   108T4983G0001 

In [141]:
parts_cost_est.to_csv('parts_cost_est.csv', index=False)
print('Done')

Done
